pkg/statistics/histogram.go

```diff
@@ -1052,38 +1052,41 @@ func (hg *Histogram) OutOfRangeRowCount(
 	}
 
-	totalPercent := leftPercent*0.5 + rightPercent*0.5
-	if totalPercent > 1 {
-		totalPercent = 1
-	}
-	rowCount = totalPercent * hg.NotNullCount()
+	totalPercent := min(leftPercent*0.5+rightPercent*0.5, 1.0)
 
-	// Upper bound logic
+	// Upper & lower bound logic.
+	upperBound := rowCount
+	if histNDV > 0 {
+		upperBound = hg.NotNullCount() / float64(histNDV)
+	}
 
 	allowUseModifyCount := sctx.GetSessionVars().GetOptObjective() != variable.OptObjectiveDeterminate
-	// Use the modifyCount as the upper bound. Note that modifyCount contains insert, delete and update. So this is
-	// a rather loose upper bound.
-	// There are some scenarios where we need to handle out-of-range estimation after both insert and delete happen.
-	// But we don't know how many increases are in the modifyCount. So we have to use this loose bound to ensure it
-	// can produce a reasonable results in this scenario.
-	if rowCount > float64(modifyCount) && allowUseModifyCount {
-		return float64(modifyCount)
-	}
-
-	// In OptObjectiveDeterminate mode, we can't rely on the modify count anymore.
-	// An upper bound is necessary to make the estimation make sense for predicates with bound on only one end, like a > 1.
-	// But it's impossible to have a reliable upper bound in all cases.
-	// We use 1/NDV here (only the Histogram part is considered) and it seems reasonable and good enough for now.
-	if !allowUseModifyCount {
-		var upperBound float64
-		if histNDV > 0 {
-			upperBound = hg.NotNullCount() / float64(histNDV)
-		}
-		if rowCount > upperBound {
-			return upperBound
-		}
+
+	if allowUseModifyCount && float64(modifyCount) > hg.NotNullCount() {
+		rowCount = totalPercent * float64(modifyCount)
+	} else {
+		rowCount = totalPercent * hg.NotNullCount()
 	}
-	return rowCount
+
+	if !allowUseModifyCount {
+		// In OptObjectiveDeterminate mode, we can't rely on the modify count anymore.
+		// An upper bound is necessary to make the estimation make sense for predicates with bound on only one end, like a > 1.
+		// We use 1/NDV here (only the Histogram part is considered) and it seems reasonable and good enough for now.
+		return min(rowCount, upperBound)
+	}
+	// If the modifyCount is large (as a percentage of the base table rows), then any out of range estimate is
+	// unreliable - because this indicates that our statistics are stale. These bounds are therefore approximations.
+	// ModifyCount counts all inserts/updates/deletes - this logic has a bias towards modifications being inserts.
+	// Assume a minimum of 1 value is found if we have a high modifyCount. This targest the situation where we search
+	// for a statisically out of range value, but inserts result in the value actually being in that modified range.
+	modifyPercent := min(float64(modifyCount)/hg.NotNullCount(), 1.0)
+	if modifyPercent == 1 || modifyPercent > totalPercent {
+		rowCount = max(rowCount, upperBound)
+	} else if rowCount < upperBound {
+		rowCount *= hg.GetIncreaseFactor(realtimeRowCount)
```

Comment:
Here is a small pitfall.
`hg.GetIncreaseFactor` can only consider the row count in the histogram. The row count in TopN and NULL is not considered. But `realtimeRowCount` counts everything.
So probably it's better to calculate the "increase factor" using `(*Index).GetIncreaseFactor()` or `(*Column).GetIncreaseFactor()` in the outside and pass it into this function.

Reply:
Corrected in latest revision